controllers/volumereplication_controller.go

```diff
@@ -209,26 +211,41 @@ func (r *VolumeReplicationReconciler) SetupWithManager(mgr ctrl.Manager, cfg *co
 }
 
 // markVolumeAsPrimary defines and runs a set of tasks required to mark a volume as primary
-func (r *VolumeReplicationReconciler) markVolumeAsPrimary(volumeID string, parameters, secrets map[string]string) (string, error) {
+func (r *VolumeReplicationReconciler) markVolumeAsPrimary(volumeID string, forcePromote bool, parameters, secrets map[string]string) (string, error) {
 	c := replication.CommonRequestParameters{
 		VolumeID:    volumeID,
 		Parameters:  parameters,
 		Secrets:     secrets,
 		Replication: r.Replication,
 	}
-	force := false
 	var markVolumeAsPrimaryTasks = []*tasks.TaskSpec{
 		{
 			Name: enableVolumeReplication,
 			Task: replication.NewEnableTask(c),
 		},
 		{
 			Name: promoteVolume,
-			Task: replication.NewPromoteVolumeTask(c, force),
+			Task: replication.NewPromoteVolumeTask(c, forcePromote),
+			KnownErrors: []codes.Code{
+				codes.FailedPrecondition,
+			},
 		},
 	}
 
-	return tasks.RunAll(markVolumeAsPrimaryTasks)
+	taskName, taskError := tasks.RunAll(markVolumeAsPrimaryTasks)
+	if taskError.Position >= 0 {
```

Comment:
why we need a Position here?

Reply:
Because we need to know which Task failed to be able to fetch it's KnownErrors.

Reply:
Take a look at https://github.com/kube-storage/volume-replication-operator/pull/42 based on the name we can check  the errors for the particular task

Reply:
That looks clean. Will rebase on top of it.